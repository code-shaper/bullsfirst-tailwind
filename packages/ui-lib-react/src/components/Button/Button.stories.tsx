import * as React from 'react';
import { Story, Meta } from '@storybook/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Card } from '../Card';
import { Button, LinkButton } from './Button';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: { onClick: { action: 'clicked' } },
} as Meta;

const Template: Story = (args) => (
  <Button
    className={args.className}
    variant={args.variant}
    color={args.color}
    size={args.size}
    disabled={args.disabled}
    onClick={args.onClick}
  >
    {args.text}
  </Button>
);

export const ButtonStory = Template.bind({});
ButtonStory.storyName = 'Button';
ButtonStory.args = {
  className: '',
  variant: 'contained',
  color: 'primary',
  size: 'medium',
  disabled: false,
  text: 'Button',
};

export const ButtonVariationsStory = () => {
  return (
    <Card className="max-w-3xl p-6">
      <h1 className="text-2xl font-medium">Buttons</h1>

      <h2 className="text-xl mt-4">Contained Buttons</h2>
      <div className="mt-1">
        <Button>Default</Button>
        <Button className="ml-2" color="primary">
          Primary
        </Button>
        <Button className="ml-2" color="secondary">
          Secondary
        </Button>
        <Button className="ml-2" disabled>
          Disabled
        </Button>
      </div>

      <h2 className="text-xl mt-6">Outlined Buttons</h2>
      <div className="mt-1">
        <Button variant="outlined">Default</Button>
        <Button className="ml-2" variant="outlined" color="primary">
          Primary
        </Button>
        <Button className="ml-2" variant="outlined" color="secondary">
          Secondary
        </Button>
        <Button className="ml-2" variant="outlined" disabled>
          Disabled
        </Button>
      </div>

      <h2 className="text-xl mt-6">Sizes</h2>
      <div className="mt-1">
        <Button variant="contained" color="primary" size="small">
          Small
        </Button>
        <Button
          className="ml-2"
          variant="contained"
          color="primary"
          size="medium"
        >
          Medium
        </Button>
        <Button
          className="ml-2"
          variant="contained"
          color="primary"
          size="large"
        >
          Large
        </Button>
      </div>
      <div className="mt-2">
        <Button variant="outlined" color="primary" size="small">
          Small
        </Button>
        <Button
          className="ml-2"
          variant="outlined"
          color="primary"
          size="medium"
        >
          Medium
        </Button>
        <Button
          className="ml-2"
          variant="outlined"
          color="primary"
          size="large"
        >
          Large
        </Button>
      </div>
    </Card>
  );
};
ButtonVariationsStory.storyName = 'Button Variations';

export const LinkButtonVariationsStory = () => {
  return (
    <Router>
      <Card className="max-w-3xl p-6">
        <h1 className="text-2xl font-medium">Link Buttons</h1>

        <h2 className="text-xl mt-4">Contained Buttons</h2>
        <div className="mt-2">
          <LinkButton to="/home">Default</LinkButton>
          <LinkButton to="/home" className="ml-2" color="primary">
            Primary
          </LinkButton>
          <LinkButton to="/home" className="ml-2" color="secondary">
            Secondary
          </LinkButton>
        </div>

        <h2 className="text-xl mt-8">Outlined Buttons</h2>
        <div className="mt-2">
          <LinkButton to="/home" variant="outlined">
            Default
          </LinkButton>
          <LinkButton
            to="/home"
            className="ml-2"
            variant="outlined"
            color="primary"
          >
            Primary
          </LinkButton>
          <LinkButton
            to="/home"
            className="ml-2"
            variant="outlined"
            color="secondary"
          >
            Secondary
          </LinkButton>
        </div>

        <h2 className="text-xl mt-8">Sizes</h2>
        <div className="mt-2">
          <LinkButton
            to="/home"
            variant="contained"
            color="primary"
            size="small"
          >
            Small
          </LinkButton>
          <LinkButton
            to="/home"
            className="ml-2"
            variant="contained"
            color="primary"
            size="medium"
          >
            Medium
          </LinkButton>
          <LinkButton
            to="/home"
            className="ml-2"
            variant="contained"
            color="primary"
            size="large"
          >
            Large
          </LinkButton>
        </div>
        <div className="mt-6">
          <LinkButton
            to="/home"
            variant="outlined"
            color="primary"
            size="small"
          >
            Small
          </LinkButton>
          <LinkButton
            to="/home"
            className="ml-2"
            variant="outlined"
            color="primary"
            size="medium"
          >
            Medium
          </LinkButton>
          <LinkButton
            to="/home"
            className="ml-2"
            variant="outlined"
            color="primary"
            size="large"
          >
            Large
          </LinkButton>
        </div>
      </Card>
    </Router>
  );
};
LinkButtonVariationsStory.storyName = 'Link Button Variations';
