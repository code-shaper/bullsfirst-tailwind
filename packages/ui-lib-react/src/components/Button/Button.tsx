import * as React from 'react';
import clsx from 'clsx';

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'contained' | 'outlined';
  color?: 'default' | 'primary' | 'secondary';
  size?: 'small' | 'medium' | 'large';
}

const baseStyles = 'font-medium rounded';

const sizeStyles = {
  small: 'text-xs px-2.5 py-1.5',
  medium: 'text-sm leading-4 px-3 py-2',
  large: 'text-sm px-4 py-2',
};

const variantStyles = {
  contained: 'border border-transparent',
  outlined: 'bg-transparent border',
};

const colorStyles = {
  contained: {
    default: 'text-neutral-900 bg-gray-200 hover:bg-gray-300',
    primary: 'text-white bg-primary-default hover:bg-primary-900',
    secondary: 'text-neutral-900 bg-secondary-default hover:bg-secondary-500',
  },
  outlined: {
    default: 'text-gray-600 border-gray-600 hover:bg-gray-200',
    primary: 'text-primary-default border-primary-default hover:bg-primary-50',
    secondary: 'text-secondary-800 border-secondary-800 hover:bg-secondary-100',
  },
};

const disabledSyles = 'cursor-default pointer-events-none';
const disabledVariantSyles = {
  contained: 'text-gray-300 bg-gray-100',
  outlined: 'text-gray-300 border-gray-300',
};

export function Button({
  className,
  variant = 'contained',
  color = 'default',
  size = 'medium',
  disabled = false,
  children,
  ...props
}: ButtonProps) {
  const buttonClass = disabled
    ? clsx(
        className,
        baseStyles,
        variantStyles[variant],
        sizeStyles[size],
        disabledSyles,
        disabledVariantSyles[variant]
      )
    : clsx(
        className,
        baseStyles,
        variantStyles[variant],
        sizeStyles[size],
        colorStyles[variant][color]
      );

  return (
    <button className={buttonClass} disabled={disabled} {...props}>
      {children}
    </button>
  );
}
