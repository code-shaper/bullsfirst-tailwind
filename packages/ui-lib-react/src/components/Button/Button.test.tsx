import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Button, LinkButton } from './Button';

describe('<Button />', () => {
  test('colors render correctly', () => {
    const { asFragment } = render(
      <div>
        <Button color="default">Submit</Button>
        <Button color="primary">Submit</Button>
        <Button color="secondary">Submit</Button>
      </div>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('class can be added via className prop', () => {
    const { asFragment } = render(<Button className="ml-2">Submit</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<LinkButton />', () => {
  test('colors render correctly', () => {
    const { asFragment } = render(
      <Router>
        <LinkButton color="default" to="/home">
          Submit
        </LinkButton>
        <LinkButton color="primary" to="/home">
          Submit
        </LinkButton>
        <LinkButton color="secondary" to="/home">
          Submit
        </LinkButton>
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('class can be added via className prop', () => {
    const { asFragment } = render(
      <Router>
        <LinkButton className="ml-2" to="/home">
          Submit
        </LinkButton>
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
