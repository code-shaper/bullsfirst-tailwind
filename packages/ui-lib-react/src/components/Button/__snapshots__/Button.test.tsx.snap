// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> class can be added via className prop 1`] = `
<DocumentFragment>
  <button
    class="ml-2 font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-neutral-900 bg-gray-200 hover:bg-gray-300"
  >
    Submit
  </button>
</DocumentFragment>
`;

exports[`<Button /> colors render correctly 1`] = `
<DocumentFragment>
  <div>
    <button
      class="font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-neutral-900 bg-gray-200 hover:bg-gray-300"
    >
      Submit
    </button>
    <button
      class="font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-white bg-primary-default hover:bg-primary-900"
    >
      Submit
    </button>
    <button
      class="font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-neutral-900 bg-secondary-default hover:bg-secondary-500"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;

exports[`<LinkButton /> class can be added via className prop 1`] = `
<DocumentFragment>
  <a
    class="ml-2 font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-neutral-900 bg-gray-200 hover:bg-gray-300"
    href="/home"
  >
    Submit
  </a>
</DocumentFragment>
`;

exports[`<LinkButton /> colors render correctly 1`] = `
<DocumentFragment>
  <a
    class="font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-neutral-900 bg-gray-200 hover:bg-gray-300"
    href="/home"
  >
    Submit
  </a>
  <a
    class="font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-white bg-primary-default hover:bg-primary-900"
    href="/home"
  >
    Submit
  </a>
  <a
    class="font-medium rounded border border-transparent text-sm leading-4 px-3 py-2 text-neutral-900 bg-secondary-default hover:bg-secondary-500"
    href="/home"
  >
    Submit
  </a>
</DocumentFragment>
`;
